pkgs: add consideration for new packages

zimbatm opened the pull request:
## Description of changes

Add a light decision-making framework to help decide what fits in the project or not. To avoid nixpkgs becoming a dump of personal projects.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

zimbatm committed:
pkgs: add consideration for new packages

Nixpkgs is not your personal depot. Add a light decision-making
framework to help decide what fits in the project or not.

mweinelt commented on pkgs/README.md:
```diff
@@ -67,7 +78,9 @@ To add a package to Nixpkgs:
-   - All [`meta`](https://nixos.org/manual/nixpkgs/stable/#chap-meta) attributes are optional, but it’s still a good idea to provide at least the `description`, `homepage` and [`license`](https://nixos.org/manual/nixpkgs/stable/#sec-meta-license).
+   - Add yourself as the maintainer of the package.
+
+   - All other [`meta`](https://nixos.org/manual/nixpkgs/stable/#chap-meta) attributes are optional, but it’s still a good idea to provide at least the `description`, `homepage` and [`license`](https://nixos.org/manual/nixpkgs/stable/#sec-meta-license).
```
I'm hesitant to characterize `license` as optional.

mweinelt commented on pkgs/README.md:
```diff
@@ -23,7 +23,18 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
+Nixpkgs. Before adding a new package, please consider the following questions:
+* Does the package have a release?
+* How realistic is it that it will be used by other people?
+* Are you going to maintain the package?
```
Wondering if we should specify a timeframe here, like

```suggestion
* Are you going to keep maintaining the package at least 6 months from today?
```

zimbatm replied:
It's a good question, I don't know.

On the surface, it seems reasonable. But it also opens up new questions like what is the appropriate duration, is there a single duration that works for all packages, and what happens if they don't respect that commitment?

Or we ask the contributor to make a good-faith commitment and leave it unspecified.

mweinelt replied:
A more reasonable timeframe would be the next NixOS release cycle, but I'm fine with a good faith commitment as well.

zimbatm replied:
It would be nice to move towards that for sure. I wonder what existing maintainers would think of this. Keeping this open for now to let it simmer a bit.

zimbatm replied:
Added some wording to make the minimum duration normative but not enforced: a0378d09042d3cc4ed5166d20c7f0f9f27913511

7c6f434c reviewed (dismissed):
For Nixpkgs that had releases _less_ stable than the average state of unstable for a few years to pretend that not having versioned releases is always a sign of instability is both disconnected from reality and hypocritical

AndersonTorres commented on pkgs/README.md:
```diff
@@ -67,7 +78,9 @@ To add a package to Nixpkgs:
-   - All [`meta`](https://nixos.org/manual/nixpkgs/stable/#chap-meta) attributes are optional, but it’s still a good idea to provide at least the `description`, `homepage` and [`license`](https://nixos.org/manual/nixpkgs/stable/#sec-meta-license).
+   - Add yourself as the maintainer of the package.
+
+   - All other [`meta`](https://nixos.org/manual/nixpkgs/stable/#chap-meta) attributes are optional, but it’s still a good idea to provide at least the `description`, `homepage` and [`license`](https://nixos.org/manual/nixpkgs/stable/#sec-meta-license).
```
platforms and mainProgram should be set, too.

zimbatm replied:
Personally I'm also expecting the license to be set, but let's have that conversation separately ok?

7c6f434c replied:
Requiring license is probably defensible from «checking it is supposed to be packageable» point of view.

infinisil replied:
This document is still a bit of a mess, there's also this section on meta which is probably a better place to change: https://github.com/NixOS/nixpkgs/blob/c5d067d0f23a3412312f036a315bd3e2a8970e9b/pkgs/README.md#meta-attributes

zimbatm replied:
I added the license question in the filters: d7b1941ca8c50ff500474bb864b634e921b41dd7

Agreed, once this is merged, I think we should re-format the doc with a focus on the contributor experience, and how they access that information.

zimbatm commented:
> For Nixpkgs that had releases less stable than the average state of unstable for a few years to pretend that not having versioned releases is always a sign of instability is both disconnected from reality and hypocritical

Was the personal attack necessary? I agree that there are some cases where packaging an unreleased version makes sense, but generally speaking, it's a good signal that the project is not ready to be used. I could change the wording to leave more room for interpretation or add some explicit exceptions.

7c6f434c commented:
> Was the personal attack necessary?

Was it really the attack? I think we should be more careful not to lose at least the implications of our own experience, as we don't always have access to interpret fully the experience of other projects.

> generally speaking, it's a good signal that the project is not ready to be used

Given that the packages I trust most to keep working are often the no-versioned-releases-ever ones, if we want to ask for some property, we should _start_ with stating the property and not just put a pretty unreliable proxy in its place.

zimbatm commented on pkgs/README.md:
```diff
@@ -23,7 +23,18 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
-To add a package to Nixpkgs:
+Each new package has a cost for the maintainers, Hydra and users downloading
+Nixpkgs. Before adding a new package, please consider the following questions:
+* Does the package have a release?
```
@7c6f434c what better metric for maturity do you propose? Let's restart the convo in this thread.

zimbatm replied:
Also what kind of packages do you have in mind? The vast majority of packages have releases.

7c6f434c replied:
I don't think there is a good universal metric, and I don't think there can be. 

If what we care about is whether the package ready for general use, we should start by asking «Is the package ready for general use?» Sure, nothing wrong with adding «For example, if the author(s) is/are working towards the first release, it is probably too early.»

Quicklisp for one example is half-full of packages that don't bother with versions and just publish commits fixing whatever small bugs are discovered.

zimbatm replied:
Thanks, that's useful. Adopted.

7c6f434c replied:
Thanks!

zimbatm committed:
tweak wording

zimbatm committed:
add license question

RaitoBezarius commented on pkgs/README.md:
```diff
@@ -23,7 +23,20 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
-To add a package to Nixpkgs:
+Each new package has a cost for the maintainers, Hydra and users downloading
+Nixpkgs. Before adding a new package, please consider the following questions:
+* Is the package ready for general use? For example, if the author(s) is/are working towards the first release, it is probably too early.
```
Remark: I will go further, I had discussions with @Foxboron on this who had some of his projects "too early packaged" and he was not really comfortable with this.

For example, we can also be explicit and say "if the author explicitly asks for not being packaged", etc. I mean, it's obvious in retrospective in packaging `unstable-XYZ`-type of packages.

I would just like to distinguish:

- software that is used and can be used because it has been there for a while, but no release or security fixes in latest HEAD, etc.
- software that is being actively developed and can be already ~usable as a sneak peek, but with absolute no quality guarantees.

The former is fine without "first release", the latter is very much meh IMHO.

7c6f434c replied:
The precedent is that we don't do much deference to upstream, though. (As in, multipled upstreams tried to stop people from packaging their announced-as-stable OSI-and-DSFG-approved-license software, and Nixpkgs ignored the requests)

Foxboron replied:
There is a difference between a maintainer going "this isn't stable software and I would urge you to wait" vs "I dont want my FOSS code published in a software repository".

I would ignore the latter if I didn't care about downstream support when encountering bugs.

zimbatm replied:
Ok, I addressed this in d5f8b87b2ece652c722a0091c793465e005f6406 . I tried compressing the decision tree so it doesn't become a full D&D adventure.

RaitoBezarius commented on pkgs/README.md:
```diff
@@ -23,7 +23,20 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
+* How realistic is it that it will be used by other people?
+* Are you going to maintain the package?
+
+If any of these questions' answer is no, then you should probably not add the
```
Offer alternatives here IMHO: "consider a NUR, etc."

zimbatm replied:
fixed in 7fae21eb99c81c2f3998db54509c6c81f6063003

RaitoBezarius commented:
Giving examples may be interesting, not sure how feasible this may be.

AndersonTorres commented on pkgs/README.md:
```diff
@@ -23,7 +23,20 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
 ## Quick Start to Adding a Package
 
-To add a package to Nixpkgs:
+Each new package has a cost for the maintainers, Hydra and users downloading
```
```suggestion
Each new package has a cost for the maintainers, Continuous Integration, caching servers and users downloading
```

More generic and user'ish.

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

zimbatm committed:
add duration recommendation

zimbatm committed:
more re-wording

zimbatm committed:
offer alternative

zimbatm commented:
Alright, I believe I have addressed all of the points. The only remaining one is the tonality point by AndersonTorres which I don't really know how to solve.

Foxboron commented:
I'm no nixpkgs contributor, but speaking from my years as Arch maintainer.

The tone is fine. Adding packages is "easy", it adds additional complexity for the distribution as a whole and removing them is "hard".

The bar to include something should always be there and people adding new packages *should* ask them self "is this only for me, or is this actually useful" before including new things.

If you add poorly maintained projects, someone down the line is going to be want to support newer compiler/dependency targets and be stuck on the poorly maintained project. This work requires effort that not everyone is.

Having a repository of all software in the world, with a complete dependency chain is an impossible job to maintain with an ever increasing amount of complexity.

AndersonTorres commented:
> The bar to include something should always be there and people adding new packages _should_ ask them self "is this only for me, or is this actually useful" before including new things.
> 

Understandable. 
On the other hand, to cite a personal example, I was the only one interested in packaging _yet another clunky Mplayer fork_. Certainly `mpv` is useful for many people, indeed there are many other people that keep it along with me now. However in the start it was merely another video player, useful only for me.

> This work requires effort that not everyone is capable of committing.

Then it should be about _accountability_. The person should be committed to keep the package first and foremost.

zimbatm commented:
There is also a catch-22 if we ask packages to be popular before they can get packaged. That's why I didn't want to write those down as hard rules.

Ultimately it's a negotiation that you have to do with yourself. What matters to me is that (1) those points are considered when doing so. And (2) that they can be used in the conversation if somebody else has a different viewpoint. That more or less we have a general understanding of what is valid to package.

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

AndersonTorres commented on pkgs/README.md:
```diff
@@ -23,7 +23,25 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
 ## Quick Start to Adding a Package
 
-To add a package to Nixpkgs:
+Each new package has a cost for the maintainers, Continuous Integration, caching servers and users downloading Nixpkgs.
```
```suggestion
We welcome new contributors of new packages to Nixpkgs, arguably the greatest software database known. However, each new package comes with a cost for the maintainers, Continuous Integration, caching servers and users downloading Nixpkgs.
```

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

zimbatm commented on pkgs/README.md:
```diff
@@ -23,7 +23,25 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
+
+Before adding a new package, please consider the following questions:
+
+* Is the package ready for general use? We don't want to include projects that are going to be abandoned immediately. In case of doubt, check with upstream.
```
These points seem close enough that they can be handled by that single point. Let's keep the list as short as possible.

RaitoBezarius commented:
> On the other hand, to cite a personal example, I was the only one interested in packaging _yet another clunky Mplayer fork_. Certainly `mpv` is useful for many people, indeed there are many other people that keep it along with me now. However in the start it was merely another video player, useful only for me.

This is not a big deal in the NixOS ecosystem, you can start as a NUR, have people be interested in it, promote it to nixpkgs when you figure out a lot of folks are interested in.

Sure, we suck at package discovery across the whole universe with NURs…

SuperSandro2000 commented on pkgs/README.md:
```diff
@@ -23,7 +23,25 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
+* How realistic is it that it will be used by other people? It's good that nixpkgs caters to various niches, but if it's a niche of 5 people it's probably too small.
+* Are you willing to maintain the package? You should care enough about the package to be willing to keep it up and running for at least one complete Nixpkgs' release life-cycle.
+
+If any of these questions' answer is no, then you should probably not add the
+package.
```
```suggestion
If any of these questions' answer is no, then you should probably not add the package.
```
big fan of one sentence per line like the rest of the file :)

SuperSandro2000 commented on pkgs/README.md:
```diff
@@ -23,7 +23,25 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
+
+This is section describes a general framework of understanding and exceptions might apply.
+
+Luckily it's pretty easy to maintain your own package set with Nix , which can then be added to the [Nix User Repository](https://github.com/nix-community/nur) project.
```
```suggestion
Luckily it's pretty easy to maintain your own package set with Nix, which can then be added to the [Nix User Repository](https://github.com/nix-community/nur) project.
```

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

AndersonTorres commented on pkgs/README.md:
```diff
@@ -23,7 +23,24 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
+
+Before adding a new package, please consider the following questions:
+
+* Is the package ready for general use? We don't want to include projects that are going to be abandoned immediately. In case of doubt, check with upstream.
```
That being said, a better wording:

```suggestion
* Is the package ready for general use? We don't want to include projects that are too imature or are going to be abandoned immediately. In case of doubt, check with upstream.
```

AndersonTorres reviewed (commented):
A non-blocking suggestion:

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

zimbatm committed:
Update pkgs/README.md

Co-authored-by: Ryan Lahfa <masterancpp@gmail.com>

AndersonTorres reviewed (approved):
_Nihil Obstat_.

infinisil reviewed (approved):
I have some cleanups for `pkgs/README.md` planned to make it similarly structured to `CONTRIBUTING.md`, with separate "How To's" (where the current Quick Start section should go) and "Conventions" sections (where the new part from this PR should go). But it's not necessary to use that new proposed structure here already, it can be moved later :)

So this looks good!

Artturin commented on pkgs/README.md:
```diff
@@ -23,7 +23,24 @@ See the [CONTRIBUTING.md](../CONTRIBUTING.md) document for more general informat
+
+* Is the package ready for general use? We don't want to include projects that are too immature or are going to be abandoned immediately. In case of doubt, check with upstream.
+* Does the project have a clear license statement? Remember that softwares are unfree by default (all rights reserved), and merely providing access to the source code does not imply its redistribution. In case of doubt, ask upstream.
+* How realistic is it that it will be used by other people? It's good that nixpkgs caters to various niches, but if it's a niche of 5 people it's probably too small.
```
The list seems to be missing a bullet point about novelty programs and programs with multiple implementations in nixpkgs already

See https://github.com/NixOS/nixpkgs/pull/258479 which adds a 4th implementation of lolcat
